src/box/box.cc

```diff
@@ -4437,6 +4454,46 @@ box_cfg(void)
 	}
 }
 
+/**
+ * Whether the recovery state triggers for a specific state have already run.
+ */
+static bool recovery_state_reached[RECOVERY_STATE_MAX];
+
+static const char *recovery_state_names[RECOVERY_STATE_MAX] = {
+	/* [RECOVERY_STATE_SNAPSHOT_RECOVERED] = */
+	"snapshot_recovered",
+	/* [RECOVERY_STATE_WAL_RECOVERED] = */
+	"wal_recovered",
+	/* [RECOVERY_STATE_INDEXES_BUILT] = */
+	"indexes_built",
+	/* [RECOVERY_STATE_SYNCED] = */
+	"synced",
+};
+
+int
+box_run_on_recovery_state(enum box_recovery_state state)
+{
+	/* Don't run triggers for the same stage twice. */
+	if (recovery_state_reached[state])
+		return 0;
+	recovery_state_reached[state] = true;
+	/*
+	 * on_recovery_state triggers expect a string, but passing a string
+	 * literal to them turns out quite ugly, because trigger_run takes
+	 * void *. So one ends up with
+	 * trigger_run(trig, (char *)"string_literal");
+	 *
+	 * Copying the string turns out less ugly and more safe, because the
+	 * resulting string becomes writable.
```

Comment:
I don't understand why we should bother at all - the trigger isn't going to modify the string, right? And `const` specifying in C is optional and doesn't really enforce constant-ness. So I see nothing wrong in writing:

```C
trigger_run(&box_on_recovery_state, (char *)recovery_state_names[state]);
```

I'm pretty sure we have a few other places where we do the same thing.

Reply:
>  the trigger isn't going to modify the string, right?

Yes, it isn't going to modify the string, and actually the string is copied anyway for Lua triggers (via `lua_pushstring` in `lbox_push_recovery_state`). So your suggestion would work just fine now.

On the other hand, the trigger run function receives a `(void *)` argument, and nothing suggests that the argument can't be written to. That's why I want to remain on the safe side for the case when someone accidentally writes to the string in a C trigger. IMO if we promise that some argument is non-const, it should really be non-const.

What do you think?

Reply:
> That's why I want to remain on the safe side for the case when someone accidentally writes to the string in a C trigger.

If someone accidentally writes to a constant string, they'll get SIGSEGV, because the section is read-only.

> IMO if we promise that some argument is non-const, it should really be non-const.

Only if it comes for free. Following the "non-const unless marked as const" rule just for the sake of it isn't worth complicating the code.

Reply:
> If someone accidentally writes to a constant string, they'll get SIGSEGV, because the section is read-only.

True. But the trigger could write to the argument conditionally. And some condition might fire on the user side only.
Anyway, I'm just inventing possible scenarios here.

> Only if it comes for free. Following the "non-const unless marked as const" rule just for the sake of it isn't worth complicating the code.

As you wish:
```diff
diff --git a/src/box/box.cc b/src/box/box.cc
index 363452f30..2bda374f4 100644
--- a/src/box/box.cc
+++ b/src/box/box.cc
@@ -4475,21 +4475,9 @@ static const char *recovery_state_names[RECOVERY_STATE_MAX] = {
 int
 box_run_on_recovery_state(enum box_recovery_state state)
 {
-       /*
-        * on_recovery_state triggers expect a string, but passing a string
-        * literal to them turns out quite ugly, because trigger_run takes
-        * void *. So one ends up with
-        * trigger_run(trig, (char *)"string_literal");
-        *
-        * Copying the string turns out less ugly and more safe, because the
-        * resulting string becomes writable.
-        */
        assert(state >= 0 && state < RECOVERY_STATE_MAX);
-       char buf[64];
-       size_t len = strlcpy(buf, recovery_state_names[state], sizeof(buf));
-       assert(len < sizeof(buf));
-       (void)len;
-       return trigger_run(&box_on_recovery_state, buf);
+       return trigger_run(&box_on_recovery_state,
+                          (char *)recovery_state_names[state]);
 }
 
 /**
```

Reply:
> True. But the trigger could write to the argument conditionally. And some condition might fire on the user side only.

All possible code paths should be covered by tests. If a code path isn't covered, anything can happen.

> As you wish

Thanks :)

Reply:
> All possible code paths should be covered by tests.

That's not what usually happens with our tests :)